### DISCUSSION TITLE
pipeline failing with cryptic error

### DISCUSSION THREAD

**Discussion by lucas-johnson** [wlandau/crew#128] [Q&A]
I'm running a large pipeline with targets and crew handling parallelization on my local machine. It's failing with a cryptic error message (see below). I was expecting the pipeline to finish successfully or at least provide more information with this error. I'm suspecting this error might OOM related, as I've had OOM issues with this pipeline previously when using future instead of crew, but those issues manifested with different error messages. Can somebody diagnose this error message?

```
▶ end pipeline [29.79 minutes]
Error:
! Error running targets::tar_make()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: invalid daemons: structure(5L, class = "errorValue")
 dispatcher running at pid 881616 with status "disk_sleep". Connection issue?
 ```
 
 I have posted a similar question on [SO](https://stackoverflow.com/questions/77188134/r-targets-pipeline-failing-with-cryptic-error)

**Comment by wlandau** [comment]
Seems related to https://github.com/ropensci/targets/discussions/1101. It might fix itself if you install the latest CRAN versions of nanonext, mirai, and crew. ~~@shikokuchuo has also mentioned that this sort of error might happen if there are enough busy threads or processes to max out the CPUs. (I am having trouble finding that thread.)~~ The improvements mentioned in https://github.com/shikokuchuo/mirai/discussions/63#discussioncomment-7062657 (development versions only right now) might also help. Other than that, it's hard to say what is happening without a minimal reprex.

**Reply by shikokuchuo** [reply]
> @shikokuchuo has also mentioned that this sort of error might happen if there are enough busy threads or processes to max out the CPUs. (I am having trouble finding that thread.) 

No I don't think that is accurate. I also can't remember if it was the case at one point, but the latest package versions would have fixed this.

**Reply by wlandau** [reply]
I was thinking back to when I got NNG-level errors when I was running `devtools::check()`, `covr::package_coverage()`, `devtools::test()`, `revdepcheck::revdep_check()`, etc. all at the same time on `crew`. You had mentioned I should just run these checks sequentially, and I am having trouble finding your exact reply. Sorry if I misrepresented it.

**Reply by shikokuchuo** [reply]
It might be the symptom of OOM killing the dispatcher process though thinking about it. Latest development mirai and crew should give you a different error message in that case.

**Reply by lucas-johnson** [reply]
Thanks for your thoughts. Certainly hear you on the reprex, just don't think I can meaningfully produce one in this context.

Sounds like we think it might be OOM killer related. Which is my hunch as well. And development versions of `crew` and `mirai` should yield different error messages? FWIW here's what i'm working with:
```
> packageVersion('crew')
[1] ‘0.5.0’
> packageVersion('mirai')
[1] ‘0.10.0’
> packageVersion('nanonext')
[1] ‘0.10.1’
```
Looks like my `nanonext` package is only slightly behind the latest CRAN version.

**Reply by lucas-johnson** [reply]
Upgrading to the latest development versions of `cran`, `mirai`, and `nanonext` yields a different error. Perhaps one that is meaningful to you? 
 
 ```
Error:
! Error running targets::tar_make()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: 'errorValue' int 11 | Incorrect state
 The mirai dispatcher is not running. Please call the start() method of the controller (e.g. your_controller$start() before using methods like push(), collect(), and scale(). If you already did, then the dispatcher may have exited too early.
 ```
 
Package versions: 
```
> packageVersion('crew')
[1] ‘0.5.0.9003’
> packageVersion('mirai')
[1] ‘0.11.0.9000’
> packageVersion('nanonext')
[1] ‘0.10.2.9000’
```

**Reply by shikokuchuo** [reply]
Hi Lucas, yes this is the new error I mentioned - 19 connection reset means it is clear that your background process has crashed (whether from OOM or something else). As we have not been able to receive anything back from it, there is unfortunately no further diagnostic. This is not just a symptom of an evaluation error (as that would be returned).

**Reply by lucas-johnson** [reply]
@shikokuchuo thanks so much for your quick response. I actually just edited my post (sorry – poor form I know –  I hadn't seen your response). I'm now unable to reproduce the error 19, and am now getting `'errorValue' int 11`. 

Checking the oom killer logs on my ubuntu server shows nothing...

**Reply by shikokuchuo** [reply]
Please disregard the 11 error - the current development versions of all the packages are not all in sync. Please use all CRAN versions + dev version of crew if you want the latest updates. I think that gets you back to error 19, which is that the process has crashed.

**Reply by lucas-johnson** [reply]
Ok will do. Thanks for your help.

**Reply by lucas-johnson** [reply]
Alright, new error. Nothing in the ubuntu OOM killer logs. 

```
Error:
! Error running targets::tar_make()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: 'errorValue' int 5 | Timed out
 dispatcher running at pid 2359593 with status "running". Connection issue?
 ```
 
 ```
> packageVersion('crew')
[1] ‘0.5.0.9003’
> packageVersion('mirai')
[1] ‘0.11.0’
> packageVersion('nanonext')
[1] ‘0.10.2’
```

**Reply by lucas-johnson** [reply]
Update - setting `memory = 'transient'` on the target that was repeatedly failing fixed the issue. This implies that the `'errorValue' int 5` was related to OOM  (or similar), but It's curious that the OS OOM killer wasn't getting invoked.

**Reply by lucas-johnson** [reply]
Back to getting hit with `'errorValue' int 5` even with transient memory. No OOM killer invocation in the logs.

**Reply by shikokuchuo** [reply]
Hi @lucas-johnson thanks for reporting. Can I suggest a 2-point plan to get to the bottom of this:

1. Use the latest dev versions of all packages - including mirai 0.11.0.9002. I've added some measures that shouldn't really be necessary, but might help in this instance.

2. If you continue getting the same error, move back to CRAN versions of all packages - nanonext, mirai and crew. This will help rule out the possibility that a recent change is causing this.

If you end up still getting the error, it may be that it really is a timeout and dispatcher is stuck elsewhere, thus unable to respond in time. The send/receives are all asynchronous, but moving the data across buffers may be the issue. In this case, we'll have to figure out a way of handling this in our packages.

**Reply by lucas-johnson** [reply]
Sorry for the long term radio silence – the error persists for 1 and 2 above ^.

I also was able to reproduce the behavior from the mirai-only example in @baslat's thread below. Perhaps  I should also try the crew-only example? Would that be helpful?

**Reply by shikokuchuo** [reply]
Hi @lucas-johnson, as these packages are in active development trying them at random points in time isn’t guaranteed to work. You are welcome with nanonext build 9005, mirai build 9010 and crew 0.6.0 to see if they solve the issue. If not, the actual error output would be helpful. Otherwise you will probably have to check back in once new CRAN releases have been made, unless you can pinpoint it down to a reproducible example.

**Reply by lucas-johnson** [reply]
I understand these packages are in active development, and I appreciate all the awesome work you and @wlandau have done. I'm sorry I haven't been more helpful given that I haven't produced a reprex. 

I've rolled back to using `future` instead of crew for now, as I have some deadlines approaching. 

I'm going to spend some more time in the interim to see if I can produce a reprex and will report back if/when I do. Thank you!

**Comment by baslat** [comment]
I'm getting a similar error, with {crew}. I'll try changing the `memory` setting.

```
  Last error: invalid daemons: structure(5L, class = "errorValue")
 dispatcher running at pid 37098 with status "running". Connection issue?
 ```

**Reply by baslat** [reply]
Changing `memory` did not fix it.

**Reply by shikokuchuo** [reply]
Hi @baslat some more details will help us to help you. Barring a reprex, the following would be useful:

1. package versions used of crew, mirai and nanonext.
2. system info - linux/mac/windows
3. does it happen every time or only on occasion
4. are you dealing with large amounts of data when this happens

**Reply by baslat** [reply]
![image](https://github.com/wlandau/crew/assets/14969998/78db4945-1903-475a-8593-05617e4c1928)

It always fails on one target (`erp_interp`), which works fine outside of the pipeline. This target produces a 6 column dataframe with 188319960 rows (188M), which takes about 2min to run when I make it manually. The below is a simplified version of my pipeline, which basically downloads yearly population data (ERP) by age, sex and geography (SA2) from the Australian Bureau of Statistics and interpolates it to daily. If you try to run it locally note I am using the [github version of {readabs}](https://github.com/mattcowgill/readabs). Not quite a reprex but hopefully it can give a clue or two!

```r
library(targets) # nolint
library(tarchetypes) # nolint
## Load your packages and functions contained in ./R

tar_source()

## 2 custom functions ----

#' Get Estimated Resident Population (ERP) data by Statistical Area Level 2 (SA2)
#'
#' This function retrieves ERP data by SA2 from the Australian Bureau of
#' Statistics (ABS) API using the readabs package, and transforms it. Strongly recommend setting
#' start_period to avoid 500 errors
#'
#' @param asgs The year of the Australian Statistical Geography Standard (ASGS) to use for the data.
#' @param start_period The start period for the data in YYYY format.
#' @param end_period The end period for the data in YYYY.
#'
#' @details
#' The function retrieves ERP data by SA2 for the specified ASGS year and time
#' period. The data is returned as a tibble with columns for sex, age, region,
#' and value. The region column contains the SA2 code and name, separated by a
#' comma.
#'
#' The function requires the readabs package to be installed and loaded. The package provides a convenient interface for retrieving data from the ABS API.
#'
#' @return A tibble containing ERP data by SA2 for the specified ASGS year and time period.
#'
#' @examples
#' # Get ERP data by SA2 for the year 2021
#' erp_2021 <- get_erp(asgs = 2021)
#'
#' # Get ERP data by SA2 for 2016 to 2018
#' erp <- get_erp(asgs = 2016, start_period = "2016", end_period = "2018")
#'
get_erp <- function(
	asgs,
	start_period = NULL,
	end_period = NULL
) {
	asgs_char <- as.character(asgs)

	asgs <- rlang::arg_match(
		asgs_char,
		c("2016", "2021")
	)
	flow <- paste0("ERP_ASGS", asgs)

	dat <- readabs::read_api(
		id = flow,
		datakey = list(
			# Male and female, not all persons
			sex = c(1, 2),
			# All age bins, but not the bin "all ages"
			age = c("A04", "A59", "A10", "A15", "A20", "A25", "A30", "A35", "A40", "A45", "A50", "A55", "A60", "A65", "A70", "A75", "A80", "8599"),
			# Just SA2s
			region_type = "SA2"
		),
		start_period = start_period,
		end_period = end_period
	)

	# transform it
	erp <- dat |>
		dplyr::transmute(
			year = time_period,
			age_band = age |>
				labelled::to_character() |>
				stringr::str_replace("-", stringr::fixed("_")) |>
				factor(levels = c("0_4", "5_9", "10_14", "15_19", "20_24", "25_29", "30_34", "35_39", "40_44", "45_49", "50_54", "55_59", "60_64", "65_69", "70+", "70_74", "75_79", "80_84", "85 and over")) |>
				forcats::fct_recode(
					`70+` = "70_74",
					`70+` = "75_79",
					`70+` = "80_84",
					`70+` = "85 and over"
				),
			gender = sex |>
				labelled::to_character() |>
				substr(1L, 1L),
			population = as.numeric(obs_value),
			dplyr::across(
				tidyselect::starts_with("asgs"),
				as.character,
				.names = "sa2"
			),
			asgs := asgs
		)
	# There are multiple rows for 70+ because of the collapse, so they need to be summed
	seventy_plus <- erp |>
		dplyr::filter(age_band == "70+") |>
		dplyr::count(year, age_band, gender, sa2, asgs, wt = population, name = "population")

	erp |>
		dplyr::filter(age_band != "70+") |>
		dplyr::bind_rows(seventy_plus) |>
		dplyr::arrange(year, sa2, age_band, gender) |>
		dplyr::mutate(erp_date = as.Date(paste0(year, "-06-30")))
}




#' Interpolate Estimated Resident Population (ERP) data to daily frequency
#'
#' This function takes yearly ERP data and interpolates it to daily frequency using linear interpolation.
#'
#' @param .data A tibble containing yearly ERP data by SA2, gender, and age band.
#'
#' @details
#' The function first expands the yearly data to daily frequency with explicit missing gaps using the `tsibble::fill_gaps()` function. It then uses linear interpolation to fill in the missing values using the `approx()` function. The resulting data is returned as a tibble with columns for SA2, gender, age band, and ERP date.
#'
#' @return A tibble containing daily ERP data by SA2, gender, age band, and ERP date.
#'
interpolate_erp <- function(.data) {
	# Expand from yearly data to daily data with explicit missing gaps
	# ~90s to make ~161M rows
	all_pops <- .data |>
		dplyr::select(any_of(c("erp_date", "sa2", "gender", "age_band", "population"))) |>
		tsibble::build_tsibble(
			index = erp_date,
			key = any_of(c("sa2", "gender", "age_band")),
			interval = tsibble::new_interval(day = 1)
		) |>
		tsibble::fill_gaps()

	# Linearly interpolate missing values
	# ~45 seconds per year
	all_pops |>
		tibble::as_tibble() |>
		# Create a row number for interpolation
		dplyr::mutate(time = dplyr::row_number()) |>
		# Interpolate missing values
		dplyr::mutate(population = approx(time, population, time)[["y"]]) |>
		dplyr::select(-time) |>
		dplyr::mutate(
			weight_date = erp_date + lubridate::years(2L),
			population = as.integer(round(population))
		) |>
		dplyr::select(
			any_of(c("sa2", "gender", "age_band")),
			erp_date,
			population
		)
}



## Targetse pipeline ----
targets::tar_option_set(
	packages = packs,
	controller = crew::crew_controller_local(workers = 8),
	error = "abridge" # styler: off
)

tar_plan(
	erp_raw = get_erp(2021, 2015, 2023),
	tar_target(
		erp_interp,
		interpolate_erp(erp_raw)
	)
) |
```

**Reply by wlandau** [reply]
This is boils down to a very simple pipeline with two targets. Might be helpful if you can isolate it to just one. Not sure if this will reproduce the error, but we will learn something if it does:

```r
tar_target(erp_interp, interpolate_erp(get_erp(2021, 2015, 2023))
```

Then we might also learn something if we use just `crew` and not `targets`:

```r
controller <- crew::crew_controller_local(workers = 1)
controller$start()
controller$push(
  command = interpolate_erp(get_erp(2021, 2015, 2023),
  globals = list(interpolate_erp = interpolate_erp, get_erp = get_erp)
)
controller$wait()
controller$pop()
controller$terminate()
```

Then if that still gives the error, it would be good to try just `mirai`:

```r
library(mirai)
daemons(n = 1, url = "ws://127.0.0.1:5700", dispatcher = TRUE)
task <- mirai(
  interpolate_erp(get_erp(2021, 2015, 2023),
  interpolate_erp = interpolate_erp,
  get_erp = get_erp,
  .signal = TRUE
)
call_mirai(task)
daemons(0)
```

**Reply by shikokuchuo** [reply]
@baslat thanks. I'll just add that it's useful for us to know that the current CRAN versions produce this error. Given the relatively small size of data involved, I am quite sure that this is not in itself the issue.

If it's possible, please let us know if you can still produce the error on current development versions of the packages (specifically using mirai 0.11.0.9002)? @wlandau currently has all of them mirrored so all you will need is:

```r
install.packages(c("nanonext", "mirai", "crew"), repos = "wlandau.r-universe.dev")
```

**Reply by baslat** [reply]
Single pipeline target fails with the same error:

```
> +1100 UTC 2023-10-12 07:09 50.41 start target erp_interp2
> +1100 UTC 2023-10-12 07:14 21.46 end pipeline [4.52 minutes]
Error:
! Error running targets::tar_make()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: invalid daemons: structure(5L, class = "errorValue")
 dispatcher running at pid 42829 with status "running". Connection issue?
```

Looks like crew alone fails too
```
> controller <- crew::crew_controller_local(workers = 1)
controller$start()
controller$push(
  command = interpolate_erp(get_erp(2021, 2015, 2023)),
  globals = list(interpolate_erp = interpolate_erp, get_erp = get_erp)
)
controller$wait()
controller$pop()
controller$terminate()
> controller$start()
> controller$push(
+   command = interpolate_erp(get_erp(2021, 2015, 2023)),
+   globals = list(interpolate_erp = interpolate_erp, get_erp = get_erp)
+ )
> controller$wait()
Error:
! invalid daemons: structure(5L, class = "errorValue")
 dispatcher running at pid 43290 with status "running". Connection issue?
> controller$pop()
NULL
> controller$terminate()
> 
```

mirai has been running for a long time and is hanging on `call_mirai(task)`

**Reply by baslat** [reply]
@shikokuchuo I updated the packages as you suggested but I still get the same error when running the one-target pipeline

![image](https://github.com/wlandau/crew/assets/14969998/b00683aa-3001-4a65-a6e9-df46668c95fd)

**Reply by shikokuchuo** [reply]
Hi @baslat sorry there was a mistake in the 'mirai' only example provided - it hasn't actually launched any workers so it will hang for you. Instead, can you try the following and let us know the output of the `status()` calls?

```r
library(mirai)
daemons(n = 1, url = "ws://127.0.0.1:0", dispatcher = TRUE)
launch_local(1)
Sys.sleep(2)
status()
task <- mirai(
  interpolate_erp(get_erp(2021, 2015, 2023),
  interpolate_erp = interpolate_erp,
  get_erp = get_erp,
  .signal = TRUE
)
status()
Sys.sleep(30)
status()
task$data
daemons(0)
```

**Reply by baslat** [reply]
Trying the updated mirai package and updated code:

![image](https://github.com/wlandau/crew/assets/14969998/15161d37-e277-4e0f-a9ee-245cfe7573a0)

**Reply by shikokuchuo** [reply]
Thanks. So we can see that the task never completes. But you are able to query the status at all times, which should be all `crew` is doing, except for some reason the status query itself is timing out in that case.

**Reply by baslat** [reply]
I'm on an M2 mac if that's relevant.

**Reply by shikokuchuo** [reply]
Thanks, I don't believe it is as your mirai-only tests confirm that your platform works as expected.

@wlandau this reads like a bug given it is deterministic and repeatable. It's puzzling as I struggle to see how you would generate a time-out in *any* case from querying the status. If there's any other way to reproduce one that might offer a clue.

**Reply by wlandau** [reply]
Do you mean a bug in `crew` itself? 

I am trying to run `interpolate_erp(get_erp(2021, 2015, 2023)` locally on my own M2 MacBook, but I get "Error: vector memory exhausted (limit reached?)"

**Reply by shikokuchuo** [reply]
I'm not sure actually. It seems it might be a memory issue and getting an error may be unavoidable - I'm just surprised to get this particular error.

Errorvalue 5 is very common and is possible everywhere there is a timeout argument in mirai, and in functions like until() in nanonext. But it is no longer the case that requesting status should generate one. I'm mentioning this in case it comes from somewhere else (I'm just assuming it's a status call - pls correct me if I'm wrong).

Other than that, if you can generate this in _any_ way in mirai I'd be interested to know, otherwise I'm going to recuse myself from this.

**Reply by wlandau** [reply]
Under the default settings in a `targets` pipeline, `crew` polls `mirai::status()` every 0.25 seconds. I tried to create my own reprex with the following example, which tries to return 4 GB of data from a task. One time it segfaulted Studio, the other two times I got "'errorValue' int 19 | Connection reset" from the task.

```r
library(mirai)
daemons(n = 1, url = "ws://127.0.0.1:0", dispatcher = TRUE)
launch_local(1)
task <- mirai(
  sample.int(n = 1e9, size = 1e9, replace = TRUE),
  .signal = TRUE
)
while (unresolved(task)) {
  print(nanonext::random(n = 4))
  print(status())
  Sys.sleep(0.25)
}
out <- task$data
lobstr::obj_size(out)
daemons(0)
```

**Reply by wlandau** [reply]
When I try to use the original code, I always get `"vector memory exhausted (limit reached?)"`.

With `mirai`:

```r
library(mirai)
daemons(n = 1, url = "ws://127.0.0.1:0", dispatcher = TRUE)
launch_local(1)
task <- mirai(
  interpolate_erp(get_erp(2021, 2015, 2023)),
  interpolate_erp = interpolate_erp,
  get_erp = get_erp,
  .signal = TRUE
)
while (unresolved(task)) {
  print(nanonext::random(n = 4))
  print(status())
  Sys.sleep(0.25)
}
ask$data
#> 'miraiError' chr Error: vector memory exhausted (limit reached?)
daemons(0)
```

With `crew`:

```r
library(crew)
controller <- crew_controller_local()
controller$start()
controller$push(
  command = interpolate_erp(get_erp(2021, 2015, 2023)),
  data = list(
    interpolate_erp = interpolate_erp,
    get_erp = get_erp
  )
)
controller$wait()
task <- controller$pop()
task$error
#> [1] "vector memory exhausted (limit reached?)"
controller$terminate()
```

**Reply by baslat** [reply]
The output of `interpolate_erp()` is about 5.7GB when I run it.

**Reply by wlandau** [reply]
@baslat, do you still run into this problem with the current versions of `nanonext`, `mirai`, and `crew` on GitHub?

**Reply by baslat** [reply]
I updated to the GitHub versions of those packages and the error persists:
![image](https://github.com/wlandau/crew/assets/14969998/be684034-ce32-4f07-88f5-92cb8b4db447)

**Reply by shikokuchuo** [reply]
@wlandau if you're looking to eliminate this error, I can offer the following analysis. Please disregard my previous comments as much has changed in the packages since.

This in my view is a real timeout. From the time the `status()` request is sent, over 10s (yes, lengthened) has elapsed. I have come to the conclusion that this is perfectly possible when returning a large enough object - could be just the serialization alone - in which case dispatcher may be blocking on that task and not be able to respond in time. To fix this, you can either bring back re-tries (can condition on errorValue being 5), or else a way for the user to resume such an interruption. The actual tasks should still be ongoing - in `mirai`, this is only a return 'errorValue' and not an actual error (does not halt execution). Hope this is clear.

**Reply by wlandau** [reply]
That makes a lot of sense. As of 1f4db3353bb0efc2a5b7443e9eaada250c83e4b8, brought back retries, and the `seconds_timeout` argument of `crew_controller_local()` now controls the overall timeout at the `crew` level. @baslat, you could try this and increase `seconds_timeout` to something large enough to cover serialization.

For large data workflows in `targets`, I am recommending that S3 buckets and/or local storage to manage data, which means `crew` and `mirai` only have to send instructions to parallel workers, not large data objects. You can make this happen with `storage = "worker" `and `retrieval = "worker"` in `targets`. Please have a look at https://books.ropensci.org/targets/performance.html#worker-storage.

**Reply by baslat** [reply]
I updated crew and mirai, and tried `tar_make()` with `seconds_timeout = 1200`, and I get a segfault error. I then updated nanonext too, and got this error:

```
Error:
! Error running targets::tar_make()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: 'errorValue' int 11 | Incorrect state
 The mirai dispatcher is not running. Please call the start() method of the controller (e.g. your_controller$start() before using methods like push(), collect(), and scale(). If you already did, then the dispatcher may have exited too early.
```

I will check out the link you sent and experiment with those settings, thanks.

**Reply by shikokuchuo** [reply]
Just to note that `'errorValue' int 11 | Incorrect state` is the 'correct' error when dispatcher in no longer running due to OOM etc. This is making sense now - think the potential issues were all fixed back in mirai 0.11.3.

**Reply by wlandau** [reply]
Yes, this seems like a likely out-of-memory error on the dispatcher. @baslat, for your purposes, `storage = "worker"` and `retrieval = "worker"` in `targets` (in `tar_target()` or `tar_option_set()`) will avoid sending all that data to the dispatcher and should solve the issue for you. Please also consider transient memory and garbage collection: https://books.ropensci.org/targets/performance.html.
